bentoml/_internal/utils/circus/watchfilesplugin.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+import os
+import typing as t
+import logging
+from typing import TYPE_CHECKING
+from pathlib import Path
+from threading import Event
+from threading import Thread
+
+import fs
+from circus.plugins import CircusPlugin
+
+from ...utils.pkg import source_locations
+from ....exceptions import MissingDependencyException
+from ...configuration import is_pypi_installed_bentoml
+from ...bento.build_config import BentoBuildConfig
+from ...bento.build_config import BentoPatternSpec
+
+try:
+    from watchfiles import watch
+except ImportError:
+    raise MissingDependencyException(
+        "watchfiles is required to use '--reload'. Install with `pip install 'bentoml[watchfiles]'`."
+    )
+
+if TYPE_CHECKING:
+    from watchfiles.main import FileChange
+
+
+logger = logging.getLogger(__name__)
+
+
+class ServiceReloaderPlugin(CircusPlugin):
+    name = "service_reloader"
+    config: dict[str, str]
+
+    def __init__(self, *args: t.Any, **config: t.Any):
+        assert "working_dir" in config, "`working_dir` is required"
+
+        super().__init__(*args, **config)
+
+        # circus/plugins/__init__.py:282 -> converts all given configs to dict[str, str]
+        self.name = self.config.get("name")
+        self.bentoml_home = self.config["bentoml_home"]
+        self.working_dir = self.config["working_dir"]
+
+        # a list of folders to watch for changes
+        watch_dirs = [self.working_dir, os.path.join(self.bentoml_home, "models")]
+
+        if not is_pypi_installed_bentoml():
+            # bentoml src from this __file__
+            logger.info(
+                "BentoML is installed via development mode, adding source root to 'watch_dirs'"
+            )
+            watch_dirs.append(t.cast(str, source_locations("bentoml")))
+
+        logger.info(f"Watching directories: {watch_dirs}")
+        self.watch_dirs = watch_dirs
+        logger.debug(f"Using {self.__class__.__qualname__} as reloader plugin.")
+
+        self.create_spec()
+
+        # a thread to restart circus
+        self.exit_event = Event()
+
+        self.file_changes = watch(
+            *self.watch_dirs,
+            watch_filter=None,
+            yield_on_timeout=True,  # NOTE: stop hanging on our tests, doesn't affect the behaviour
+            stop_event=self.exit_event,
+            rust_timeout=0,  # NOTE: we should set reload_delay to 0 to avoid unnecessary logs
+        )
```

Comment:
What is the performance issues of watchfiles if the watch set is very large?

Reply:
i haven't explored that part just yet.

Reply:
There are none I believe.